base/applications/mspaint/atlimagedx.h

```diff
@@ -105,37 +83,59 @@ class CImageDx : public CImage
         if (!GetClsidFromFileType(&clsid, FileType))
             return E_FAIL;
 
+        if (!GetFN(CreateBitmapFromHBITMAP, "GdipCreateBitmapFromHBITMAP") ||
+            !GetFN(BitmapSetResolution, "GdipBitmapSetResolution") ||
+            !GetFN(SaveImageToFile, "GdipSaveImageToFile") ||
+            !GetFN(DisposeImage, "GdipDisposeImage"))
+        {
+            return E_FAIL;
+        }
+
         // create a GpBitmap from HBITMAP
         GpBitmap *pBitmap = NULL;
-        GetCommon().CreateBitmapFromHBITMAP(m_hbm, NULL, &pBitmap);
+        CreateBitmapFromHBITMAP(m_hbm, NULL, &pBitmap);
 
         // set the resolution
-        SetResolution(pBitmap, xDpi, yDpi);
+        BitmapSetResolution(pBitmap, xDpi, yDpi);
 
         // save to file
-        Status status;
-        status = GetCommon().SaveImageToFile(pBitmap, pszNameW, &clsid, NULL);
+        Status status = SaveImageToFile(pBitmap, pszNameW, &clsid, NULL);
 
         // destroy GpBitmap
-        GetCommon().DisposeImage(pBitmap);
+        DisposeImage(pBitmap);
 
         return (status == Ok ? S_OK : E_FAIL);
     }
 
 protected:
     // get procedure address of the DLL
-    template <typename TYPE>
-    TYPE AddrOf(const char *name) const
+    template <typename FN_T>
+    bool GetFN(FN_T& fn, const char *name)
     {
-        FARPROC proc = ::GetProcAddress(GetCommon().hinstGdiPlus, name);
-        return reinterpret_cast<TYPE>(proc);
+        if (fn)
+            return true;
+        HINSTANCE hGdiplus = ::GetModuleHandleW(L"gdiplus.dll");
+        fn = reinterpret_cast<FN_T>(::GetProcAddress(hGdiplus, name));
+        return fn != NULL;
     }
 
-    typedef St (WINGDIPAPI *GETIMAGEHORIZONTALRESOLUTION)(Im *, float*);
-    typedef St (WINGDIPAPI *GETIMAGEVERTICALRESOLUTION)(Im *, float*);
-    typedef St (WINGDIPAPI *BITMAPSETRESOLUTION)(Bm *, float, float);
-
-    GETIMAGEHORIZONTALRESOLUTION    GetImageHorizontalResolution;
-    GETIMAGEVERTICALRESOLUTION      GetImageVerticalResolution;
-    mutable BITMAPSETRESOLUTION     BitmapSetResolution;
+    typedef decltype(&Gdiplus::DllExports::GdipGetImageHorizontalResolution)
+        FN_GetImageHorizontalResolution;
+    typedef decltype(&Gdiplus::DllExports::GdipGetImageVerticalResolution)
+        FN_GetImageVerticalResolution;
+    typedef decltype(&Gdiplus::DllExports::GdipBitmapSetResolution) FN_BitmapSetResolution;
+    typedef decltype(&Gdiplus::DllExports::GdipCreateBitmapFromHBITMAP) FN_CreateBitmapFromHBITMAP;
+    typedef decltype(&Gdiplus::DllExports::GdipCreateBitmapFromFile) FN_CreateBitmapFromFile;
+    typedef decltype(&Gdiplus::DllExports::GdipCreateHBITMAPFromBitmap) FN_CreateHBITMAPFromBitmap;
+    typedef decltype(&Gdiplus::DllExports::GdipSaveImageToFile) FN_SaveImageToFile;
+    typedef decltype(&Gdiplus::DllExports::GdipDisposeImage) FN_DisposeImage;
```

Comment:
So now, those cannot be used from the CImage class? (or maybe because they were private?)

Reply:
I should respect compatibility of ATL headers between MS ATL and ReactOS ATL.